### PR TITLE
posts: Test Best Practices blog

### PR DESCRIPTION
**Pull request by EdDev**

### CODE REVIEW

File: _posts/2020-10-19-tests-best-practices.md [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+traceback) to ease debugging.
+
+### Shared Resources
+Shared resources between tests have the potential to leak state changes
+between tests, violating the test isolation.
+
+Therefore, whenever possible it would be always preferable to recreate any
+resource from scratch for each and every test.
+
+Said that, one cannot ignore that some tests (e.g. e2e) pay a heavy cost
```
Comment (RamLavi) [inline _posts/2020-10-19-tests-best-practices.md]: `s/Said that,/That Said,` or `s/Said that,/Having said that,`

File: _posts/2020-10-19-tests-best-practices.md [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+
+In such scenarios, a more pragmatic approach is in order.
+Sharing resources across a group of tests with an immutable restriction many
+times worth the management effort and results with faster and less costly
+tests.
+
+### Continue On Failure
+Expect tests to fail, stopping on the first failure may not be informative
+enough.
+Said that, leave behind enough information to debug and troubleshoot the
```
Comment (RamLavi) [inline _posts/2020-10-19-tests-best-practices.md]: ditto

**Review by RamLavi — CHANGES_REQUESTED** [review]
added some grammer nits, plus, I would add simple code examples to the following topics: 
* Fail First
* Test body and Fixture distinction - the example is good but will be much more readable if you add code example.
* Workarounds

**Review by oshoval — COMMENTED** [review]
Missing description, even couple words please.

Maybe worth to add what is the suggested practice on 2nd assert, if its ok that they happen silently or we should write such they wont happen

for example if there is a ptr in the Context, 
a BeforeEach that should set it, unless the BeforeEach has an exception itself,
a AfterEach that deref the ptr.

1. in case the BeforeEach fails on the first test, the ptr might be nil and there will be a silect panic in the AfterEach
2. in case the previous test for example had a AfterEach that was done fine,
but has Expect delete vm success, and didnt nil the vm ptr, and on the next test the ptr wasnt set to a new vm, because the BeforeEach failed, then the AfterEach will fail silent on the Expect of delete vm

**Comment by EdDev** [comment]
> Missing description, even couple words please.
> 
> Maybe worth to add what is the suggested practice on 2nd assert, if its ok that they happen silently or we should write such they wont happen
> 
> for example if there is a ptr in the Context,
> a BeforeEach that should set it, unless the BeforeEach has an exception itself,
> a AfterEach that deref the ptr.
> 
>     1. in case the BeforeEach fails on the first test, the ptr might be nil and there will be a silect panic in the AfterEach
> 
>     2. in case the previous test for example had a AfterEach that was done fine,
>        but has Expect delete vm success, and didnt nil the vm ptr, and on the next test the ptr wasnt set to a new vm, because the BeforeEach failed, then the AfterEach will fail silent on the Expect of delete vm

This is an implementation low level detail that relates to Golang and ginkgo specifically, this post is about language and test framework agnostic practices.
If you manage to generalize the point you raised to a concept, then it may fit., But as I see it now, it is too specific to an implementation detail.

File: _posts/2020-10-19-tests-best-practices.md [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+Indirect operations that are not in the main focus of the test, better be
+placed in the setup or teardown fixtures and not in the test body.
+
+As an example, testing if an entity can be successfully created fits a test
+body.
+However, testing if an entity can communicate with another is about
+the communication check and not its creation or deletion, therefore, the
+creation will better fit the setup (and deletion into the teardown) while the
+communication check fits the test body.
+
```
Comment (RamLavi) [inline _posts/2020-10-19-tests-best-practices.md:109]: moving your verbal examle to pseudo code example
example (1): entity creation test
```
Test setup:

Test body:
create_entity()
check_entity_creation_success()
```

example (2): entity communication test
```
Test setup:
create_entity()
check_entity_creation_success()

Test body:
check_entity_connectivity()
```

File: _posts/2020-10-19-tests-best-practices.md [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+  troubleshooting failures.
+
+Therefore, an amount of uncertainty is introduced, something that tests try to
+avoid in order to be reproducible and consistent in their reported results.
+
+Projects should balance between efficient runtime, covering concurrent access
+to the SUT and the risk of uncertainty and debugging effort.
+
+
+## Workarounds
```
Comment (RamLavi) [inline _posts/2020-10-19-tests-best-practices.md:219]: in this section you mention parallel tests, which make me wonder if this section should be s sub-section of `## Parallel Tests`? if it is, then maybe change the the title to `### Workarounds`.

File: _posts/2020-10-19-tests-best-practices.md [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+This section is about patterns that attempt to solve needs in an questionable
```
Comment (RamLavi) [inline _posts/2020-10-19-tests-best-practices.md]:
```suggestion
This section is about patterns that attempt to solve needs in a questionable
```

File: _posts/2020-10-19-tests-best-practices.md (hunk 4 above)
Comment (RamLavi) [inline _posts/2020-10-19-tests-best-practices.md:219]: I'm having trouble picturing these `questionable manner` patterns from the text.. if not an example, can you elaborate on it?

File: _posts/2020-10-19-tests-best-practices.md (hunk 1 above)
Comment (EdDev) [reply _posts/2020-10-19-tests-best-practices.md]: Done, thanks!

File: _posts/2020-10-19-tests-best-practices.md (hunk 2 above)
Comment (EdDev) [reply _posts/2020-10-19-tests-best-practices.md]: Done, thanks!

File: _posts/2020-10-19-tests-best-practices.md (hunk 4 above)
Comment (EdDev) [reply _posts/2020-10-19-tests-best-practices.md:219]: I see parallel tests as somewhere in between a sensible optimization and a risk due to the concurrency introduced.
Therefore, I placed it by its own, not explicitly recommended and not discouraged.

Not sure if that is convincing enough or good.

File: _posts/2020-10-19-tests-best-practices.md (hunk 5 above)
Comment (EdDev) [reply _posts/2020-10-19-tests-best-practices.md]: Done, thanks!

File: _posts/2020-10-19-tests-best-practices.md (hunk 4 above)
Comment (EdDev) [reply _posts/2020-10-19-tests-best-practices.md:219]: These are the "Stop On Failure" and "Clean on Test Setup", I am not clear what exactly needs to be clarified or how.

File: _posts/2020-10-19-tests-best-practices.md (hunk 3 above)
Comment (EdDev) [reply _posts/2020-10-19-tests-best-practices.md:109]: Got it, thanks!

**Pull request merged**